### PR TITLE
Add Decimal256 to `ScalarValue`

### LINKED ISSUE #7049: Add Decimal256 to ScalaValue
### Is your feature request related to a problem or challenge?

For Decimal256 support, we need Decimal256 to be in ScalarValue.

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #7049.

Related: #7046.

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/common/src/scalar.rs [hunk 1]
```diff
@@ -1038,6 +1054,7 @@ macro_rules! impl_op_arithmetic {
                 get_sign!($OPERATION),
                 true,
             )))),
+            // todo: Add Decimal256 support
```
Comment (viirya) [inline datafusion/common/src/scalar.rs:1057]: The error of running

```
create table t as values (arrow_cast(123, 'Decimal256(5,2)'));
select AVG(column1) from t;
```

in decimal.slt comes from lack of implementation of Decimal256 Scalar arithmetic op.

It seems to be not easy to finish right now. So I left it as todo here.

File: datafusion/core/tests/sqllogictests/test_files/arrow_typeof.slt [hunk 2]
```diff
@@ -180,23 +180,29 @@ drop table foo
 
 statement ok
 create table foo as select
-  arrow_cast(100, 'Decimal128(5,2)') as col_d128
-  -- Can't make a decimal 156:
-  -- This feature is not implemented: Can't create a scalar from array of type "Decimal256(3, 2)"
-  --arrow_cast(100, 'Decimal256(5,2)') as col_d256
+  arrow_cast(100, 'Decimal128(5,2)') as col_d128,
+  arrow_cast(100, 'Decimal256(5,2)') as col_d256
```
Comment (viirya) [inline datafusion/core/tests/sqllogictests/test_files/arrow_typeof.slt:184]: Now it is okay to create table using decimal256.

File: datafusion/core/tests/sqllogictests/test_files/arrow_typeof.slt [hunk 3]
```diff
@@ -180,23 +180,29 @@ drop table foo
...
+Decimal128(5, 2) Decimal256(5, 2)
+
 
+query RR
+SELECT
+  col_d128,
+  col_d256
+  FROM foo;
+----
+100 100.00
```
Comment (viirya) [inline datafusion/core/tests/sqllogictests/test_files/arrow_typeof.slt:205]: Though, I'm not sure why displayed value for decimal256 shows decimal points but decimal128 doesn't.

Needs to dig into where is the code for displaying it. But for now it seems minor to me. So just leave as is.

**Comment by viirya** [comment]
Oh, seems CI won't regen protobuf code?

**Review by Dandandan — APPROVED** [review]

File: datafusion/expr/src/type_coercion/aggregates.rs [hunk 4]
```diff
@@ -441,6 +455,11 @@ pub fn avg_sum_type(arg_type: &DataType) -> Result<DataType> {
             let new_precision = DECIMAL128_MAX_PRECISION.min(*precision + 10);
             Ok(DataType::Decimal128(new_precision, *scale))
         }
+        DataType::Decimal256(precision, scale) => {
+            // in the spark, the sum type of avg is DECIMAL(min(38,precision+10), s)
```
Comment (Dandandan) [inline datafusion/expr/src/type_coercion/aggregates.rs:459]:
```suggestion
            // in Spark the sum type of avg is DECIMAL(min(38,precision+10), s)
```

**Comment by viirya** [comment]
Hmm, seems the failed tests in CI are unrelated. They are failed too on main branch as I run them locally

**Comment by alamb** [comment]
I merged up from main to get the fix for https://github.com/apache/arrow-datafusion/issues/7078

I plan to merge this PR when CI passes

**Comment by viirya** [comment]
Thank you @Dandandan @alamb

**Pull request merged**